hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/subscription/SubscriptionHealthCalculator.java

```diff
@@ -0,0 +1,139 @@
+import pl.allegro.tech.hermes.api.Subscription;
+import pl.allegro.tech.hermes.api.SubscriptionHealth;
+import pl.allegro.tech.hermes.api.SubscriptionHealth.Problem;
+
+import java.util.EnumSet;
+
+import static java.lang.Double.parseDouble;
+import static java.util.EnumSet.noneOf;
+import static pl.allegro.tech.hermes.api.Subscription.State.SUSPENDED;
+import static pl.allegro.tech.hermes.api.SubscriptionHealth.Problem.LAGGING;
+import static pl.allegro.tech.hermes.api.SubscriptionHealth.Problem.MALFUNCTIONING;
+import static pl.allegro.tech.hermes.api.SubscriptionHealth.Problem.RECEIVING_MALFORMED_MESSAGES;
+import static pl.allegro.tech.hermes.api.SubscriptionHealth.Problem.SLOW;
+import static pl.allegro.tech.hermes.api.SubscriptionHealth.Problem.TIMING_OUT;
+import static pl.allegro.tech.hermes.api.SubscriptionHealth.Problem.UNREACHABLE;
+
+@Component
+public class SubscriptionHealthCalculator {
+    private static final int MAX_LAG_IN_SECONDS = 600;
+    private static final double MIN_SUBSCRIPTION_TO_TOPIC_SPEED_RATIO = 0.8;
+    private static final double MAX_OTHER_ERRORS_RATIO = 0.5;
+    private static final double MAX_TIMEOUTS_RATIO = 0.1;
+    private static final double MAX_5XX_ERRORS_RATIO = 0.1;
+    private static final double MAX_4XX_ERRORS_RATIO = 0.1;
+
+    public SubscriptionHealth calculateHealth(Subscription subscription, pl.allegro.tech.hermes.api.TopicMetrics topicMetrics,
+                                              pl.allegro.tech.hermes.api.SubscriptionMetrics subscriptionMetrics) {
+        if (isSuspended(subscription)) {
+            return SubscriptionHealth.HEALTHY;
+        } else {
+            return calculateActiveSubscriptionHealth(subscription, topicMetrics, subscriptionMetrics);
+        }
+    }
+
+    private boolean isSuspended(Subscription subscription) {
+        return subscription.getState() == SUSPENDED;
+    }
+
+    private SubscriptionHealth calculateActiveSubscriptionHealth(Subscription subscription, pl.allegro.tech.hermes.api.TopicMetrics topicMetrics,
+                                                                 pl.allegro.tech.hermes.api.SubscriptionMetrics subscriptionMetrics) {
+        try {
+            TopicMetrics internalTopicMetrics = new TopicMetrics(topicMetrics);
+            SubscriptionMetrics internalSubscriptionMetrics = new SubscriptionMetrics(subscriptionMetrics);
+            EnumSet<Problem> healthProblems = calculateHealthProblems(subscription, internalTopicMetrics, internalSubscriptionMetrics);
+            return SubscriptionHealth.of(healthProblems);
+        } catch (NumberFormatException e) {
+            return SubscriptionHealth.NO_DATA;
+        }
+    }
+
+    private EnumSet<Problem> calculateHealthProblems(Subscription subscription, TopicMetrics topicMetrics, SubscriptionMetrics subscriptionMetrics) {
+        EnumSet<Problem> healthProblems = noneOf(Problem.class);
+        if (isLagging(topicMetrics, subscriptionMetrics)) {
```

Comment:
You could add more behaviour to Problem instances so that each Problem can detect itself based on given Subscription (since this is SubscriptionHealthCalculator). Then you could iterate over entire class extent and make those problems raise themselves :-)